azure/package_pool.go

```diff
@@ -0,0 +1,240 @@
+	if progress != nil {
+		progress.InitBar(0, false, aptly.BarGeneralBuildFileList)
+		defer progress.ShutdownBar()
+	}
+
+	paths, _, err := pool.az.internalFilelist("", progress)
+	return paths, err
+}
+
+func (pool *PackagePool) LegacyPath(filename string, checksums *utils.ChecksumInfo) (string, error) {
+	return "", errors.New("Azure package pool does not support legacy paths")
+}
+
+func (pool *PackagePool) Size(path string) (int64, error) {
+	blob := pool.az.blobURL(path)
+	props, err := blob.GetProperties(context.Background(), azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return 0, errors.Wrapf(err, "error examining %s from %s", path, pool)
+	}
+
+	return props.ContentLength(), nil
+}
+
+func (pool *PackagePool) Open(path string) (aptly.ReadSeekerCloser, error) {
+	blob := pool.az.blobURL(path)
+
+	temp, err := ioutil.TempFile("", "blob-download")
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temporary file for blob download")
+	}
+
+	defer os.Remove(temp.Name())
+
+	err = azblob.DownloadBlobToFile(context.Background(), blob, 0, 0, temp, azblob.DownloadFromBlobOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "error downloading blob at %s", path)
+	}
+
+	return temp, nil
+}
+
+func (pool *PackagePool) Remove(path string) (int64, error) {
+	const leaseDuration = 30
+
+	blob := pool.az.blobURL(path)
+	lease, err := blob.AcquireLease(context.Background(), "", leaseDuration, azblob.ModifiedAccessConditions{})
+	if err != nil {
+		return 0, errors.Wrapf(err, "error acquiring lease on %s from %s", path, pool)
+	}
+
+	defer blob.BreakLease(context.Background(), 0, azblob.ModifiedAccessConditions{})
+
+	leasedAccessConditions := azblob.BlobAccessConditions{
+		LeaseAccessConditions: azblob.LeaseAccessConditions{LeaseID: lease.LeaseID()},
+	}
+
+	props, err := blob.GetProperties(context.Background(), leasedAccessConditions, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return 0, errors.Wrapf(err, "error getting props of %s from %s", path, pool)
+	}
+
+	_, err = blob.Delete(context.Background(), azblob.DeleteSnapshotsOptionNone, leasedAccessConditions)
+	if err != nil {
+		return 0, errors.Wrapf(err, "error deleting %s from %s", path, pool)
+	}
+
+	return props.ContentLength(), nil
+}
+
+func (pool *PackagePool) Import(srcPath, basename string, checksums *utils.ChecksumInfo, move bool, checksumStorage aptly.ChecksumStorage) (string, error) {
+	const leaseDuration = 30
+
+	if checksums.MD5 == "" || checksums.SHA256 == "" || checksums.SHA512 == "" {
+		// need to update checksums, MD5 and SHA256 should be always defined
+		var err error
+		*checksums, err = utils.ChecksumsForFile(srcPath)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	path := pool.buildPoolPath(basename, checksums)
+	blob := pool.az.blobURL(path)
+
+	var leaseID string
+	lease, err := blob.AcquireLease(context.Background(), "", leaseDuration, azblob.ModifiedAccessConditions{})
```

Comment:
Is the lease strictly needed? Without periodically renewing the lease, this will set the timeout of the operation to the lease duration from my understanding.

Reply:
Hmm, it was intended to serve as an "operation lock" of sorts...but I think I was indeed over-eager with the use of leases, since many of this wouldn't have particularly bad behavior if run in parallel (e.g. the same file gets uploaded twice, but Azure won't let the uploads corrupt each other already, so it wouldn't be an issue). I'll remove the lease usage (probably some other areas in the file I can remove it from as well).